File: bindings/rust/src/kinship.rs
//! The kinship family: seven verbs, one question shape.
//!
//! `similar · dups · clusters · echoes · concepts · fragments · distinct` all
//! ask "what resembles what, and how much is that worth?", so they share one
//! `[analytic.params]` struct and one builder. Which verb you picked decides the
//! row schema and how the corpus is walked; every axis below —
//! [`channel`](Kinship::channel), [`unit`](Kinship::unit), the admission
//! threshold, [`min_grade`](Kinship::min_grade) — means the same thing on all of
//! them, which is the property `relate --help` calls "one vocabulary, every
//! verb".
//!
//! The wire keeps seven ops because `[analytic.verbs]` is append-only, but the
//! `relate` binary has since folded five of them into two verbs plus a `--shape`
//! axis. [`Form`] is where that translation lives, and it is the only place in
//! the crate that knows the CLI spelling differs from the contract's.

use std::path::PathBuf;

use irregex::contract::schema::VERBS;
use irregex::contract::{Channel, Grade, Polarity, Unit};
use irregex::runtime::relay::{Bin, Invocation};
use irregex::runtime::{Error, Query, Result, Rows, Wire, answer, struct_size, sys};

/// How one contract op is spelled by today's `relate`, and which defaults the
/// verb implies. `shape` is the axis the fold introduced; `thresholds` is false
/// for the one verb (`similar`) that ranks rather than admits, and so takes no
/// cutoff flag.
struct Form {
    cli: &'static str,
    shape: Option<&'static str>,
    channel: Channel,
    unit: Unit,
    thresholds: bool,
}

const fn form(op: u32) -> Form {
    let (cli, shape, channel, unit, thresholds) = match op {
        super::OP_DUPS => ("echoes", Some("pairs"), Channel::Copies, Unit::File, true),
        super::OP_CLUSTERS => (
            "echoes",
            Some("families"),
            Channel::Copies,
            Unit::File,
            true,
        ),
        super::OP_ECHOES => ("echoes", Some("pairs"), Channel::Twins, Unit::File, true),
        super::OP_CONCEPTS => (
            "echoes",
            Some("families"),
            Channel::Shapes,
            Unit::Function,
            true,
        ),
        super::OP_FRAGMENTS => (
            "echoes",
            Some("families"),
            Channel::Twins,
            Unit::Function,
            true,
        ),
        super::OP_DISTINCT => (
            "echoes",
            Some("distinct"),
            Channel::Copies,
            Unit::File,
            true,
        ),
        // OP_SIMILAR, and any op a newer contract adds to this family.
        _ => ("similar", None, Channel::Copies, Unit::File, false),
    };
    Form {
        cli,
        shape,
        channel,
        unit,
        thresholds,
    }
}

/// A kinship question, built axis by axis and asked with [`rows`](Self::rows).
///
/// Nothing runs until then, and the same request answers through whichever tier
/// is available — in-process when the analytic plane is present, the `relate`
/// binary otherwise, with identical rows either way.
#[derive(Debug, Clone)]
pub struct Kinship {
    op: u32,
    /// A file path, `path#Lnnn`, or probe text; absent = the corpus-wide sweep.
    target: Option<String>,
    roots: Vec<PathBuf>,
    matching: Vec<String>,
    channel: Channel,
    unit: Unit,
    max_distance: Option<f64>,
    min_echo: Option<f64>,
    min_grade: Grade,
    min_size: u32,
    min_lines: u32,
    top: u32,
    no_index: bool,
}

impl Kinship {
    /// One probe against the corpus.
    pub(crate) fn targeted(op: u32, target: String) -> Self {
        Self::seeded(op, Some(target))
    }

    /// The corpus against itself.
    pub(crate) fn sweeping(op: u32) -> Self {
        Self::seeded(op, None)
    }

    fn seeded(op: u32, target: Option<String>) -> Self {
        let f = form(op);
        Self {
            op,
            target,
            roots: Vec::new(),
            matching: Vec::new(),
            channel: f.channel,
            unit: f.unit,
            max_distance: None,
            min_echo: None,
            min_grade: Grade::None,
            min_size: 0,
            min_lines: 0,
            top: 0,
            no_index: false,
        }
    }

    /// Scope the corpus. No roots = the CWD walk, exactly as a bare CLI run.
    #[must_use]
    pub fn root(mut self, path: impl Into<PathBuf>) -> Self {
        self.roots.push(path.into());
        self
    }

    /// Scope the corpus to several roots at once.
    #[must_use]
    pub fn roots<I: IntoIterator<Item = P>, P: Into<PathBuf>>(mut self, paths: I) -> Self {
        self.roots.extend(paths.into_iter().map(Into::into));
        self
    }

    /// Price kinship only inside the files an exact pattern admits.
    /// Repeatable; see [`blast`] for the composed verbs built on it.
    #[must_use]
    pub fn matching(mut self, pattern: impl Into<String>) -> Self {
        self.matching.push(pattern.into());
        self
    }

    /// Which kinship signal to read (`--as`). Also selects which admission
    /// threshold applies — see [`Channel::polarity`].
    #[must_use]
    pub fn channel(mut self, channel: Channel) -> Self {
        self.channel = channel;
        self
    }

    /// What a row *is*: a whole file, a function, or a single match.
    #[must_use]
    pub fn unit(mut self, unit: Unit) -> Self {
        self.unit = unit;
        self
    }

    /// Admit a candidate at distance ≤ `t` (the distance-polarity channels).
    #[must_use]
    pub fn max_distance(mut self, t: f64) -> Self {
        self.max_distance = Some(t);
        self
    }

    /// Admit a candidate whose bytes−structure gap is ≥ `e` (the `twins` channel).
    #[must_use]
    pub fn min_echo(mut self, e: f64) -> Self {
        self.min_echo = Some(e);
        self
    }

    /// Withhold anything weaker than `grade`. Empty beats background.
    #[must_use]
    pub fn min_grade(mut self, grade: Grade) -> Self {
        self.min_grade = grade;
        self
    }

    /// Smallest family worth reporting.
    #[must_use]
    pub fn min_size(mut self, n: u32) -> Self {
        self.min_size = n;
        self
    }

    /// Smallest fragment worth comparing — the sub-file noise floor.
    #[must_use]
    pub fn min_lines(mut self, n: u32) -> Self {
        self.min_lines = n;
        self
    }

    /// Cap the answer. `0` = the engine's default.
    #[must_use]
    pub fn top(mut self, n: u32) -> Self {
        self.top = n;
        self
    }

    /// Force the live build, skipping the persisted atlas. Acceleration never
    /// changes the answer, so this is a diagnostic lever, not a correctness one.
    #[must_use]
    pub fn no_index(mut self) -> Self {
        self.no_index = true;
        self
    }

    /// Ask the question.
    ///
    /// # Errors
    /// [`Error::Unrepresentable`] when a threshold contradicts the channel's
    /// polarity; [`Error::SchemaDrift`] when a loaded library's row tables
    /// disagree with this build; otherwise the usual spawn / engine failures.
    pub fn rows(&self) -> Result<Rows> {
        // Fail closed rather than send a threshold this channel ignores: a
        // silently dropped `--max-distance` on `twins` reads as an unfiltered
        // answer, which is the one failure mode calibration exists to prevent.
        match (self.channel.polarity(), self.max_distance, self.min_echo) {
            (Polarity::Gap, Some(_), _) | (Polarity::Distance, _, Some(_)) => {
                return Err(Error::Unrepresentable(format!(
                    "channel `{}` admits on {}, not the other threshold",
                    self.channel,
                    self.channel.admits()
                )));
            },
            _ => {},
        }
        answer(self)
    }

    /// The threshold flag this channel admits on, paired with its value.
    fn threshold(&self) -> Option<(&'static str, f64)> {
        match self.channel.polarity() {
            Polarity::Gap => self.min_echo.map(|e| ("--min-echo", e)),
            Polarity::Distance => self.max_distance.map(|t| ("--max-distance", t)),
        }
    }
}

impl Query for Kinship {
    fn op(&self) -> u32 {
        self.op
    }

    fn roots(&self) -> &[PathBuf] {
        &self.roots
    }

    fn wire(&self) -> Wire<'_> {
        let target = self.target.as_deref().unwrap_or_default().as_bytes();
        let mut flags = 0;
        if self.max_distance.is_some() {
            flags |= sys::AN_MAX_DISTANCE;
        }
        if self.min_echo.is_some() {
            flags |= sys::AN_MIN_ECHO;
        }
        if self.no_index {
            flags |= sys::AN_NO_INDEX;
        }
        if self.op == super::OP_DISTINCT {
            flags |= sys::AN_DISTINCT;
        }
        Wire::Kinship(
            sys::KinshipParams {
                struct_size: struct_size::<sys::KinshipParams>(),
                flags,
                target: if self.target.is_some() {
                    target.as_ptr()
                } else {
                    std::ptr::null()
                },
                target_len: target.len(),
                channel: self.channel.ordinal(),
                unit: self.unit.ordinal(),
                max_distance: self.max_distance.unwrap_or_default(),
                min_echo: self.min_echo.unwrap_or_default(),
                min_grade: self.min_grade.ordinal(),
                min_size: self.min_size,
                min_lines: self.min_lines,
                top: self.top,
            },
            std::marker::PhantomData,
        )
    }

    fn argv(&self) -> Result<Invocation> {
        let f = form(self.op);
        let schema = VERBS
            .get(self.op.saturating_sub(1) as usize)
            .map_or(0, |v| v.schema);
        let mut args = vec![f.cli.to_owned()];
        if let Some(t) = &self.target {
            args.push(t.clone());
        }
        args.extend(["--as".to_owned(), self.channel.as_str().to_owned()]);
        args.extend(["--unit".to_owned(), self.unit.as_str().to_owned()]);
        if let Some(shape) = f.shape {
            args.extend(["--shape".to_owned(), shape.to_owned()]);
        }
        for pattern in &self.matching {
            args.extend(["--matching".to_owned(), pattern.clone()]);
        }
        if f.thresholds {
            if let Some((flag, value)) = self.threshold() {
                args.extend([flag.to_owned(), format!("{value}")]);
            }
            if self.min_size > 0 {
                args.extend(["--min-size".to_owned(), self.min_size.to_string()]);
            }
            if self.min_lines > 0 {
                args.extend(["--min-lines".to_owned(), self.min_lines.to_string()]);
            }
        }
        if self.min_grade != Grade::None {
            args.extend(["--min-grade".to_owned(), self.min_grade.as_str().to_owned()]);
        }
        if self.top > 0 {
            args.extend(["--top".to_owned(), self.top.to_string()]);
        }
        if self.no_index {
            args.push("--no-index".to_owned());
        }
        args.push("--json".to_owned());
        args.extend(self.roots.iter().map(|p| p.display().to_string()));
        Ok(Invocation::json(Bin::Relate, schema, args))
    }
}

File: bindings/go/kinship.go
// Package relate is the compression plane: the questions a regex cannot ask.
// What resembles this file, what repeats in this corpus, which files explain this
// text most cheaply, and where did this pasted snippet come from.
//
// Every verb here scores kinship by how cheaply one thing describes another, and
// every score arrives banded ([analytic.Grade]) so background never reads as a
// hit: an answer of five strangers at distance 0.78 says so instead of looking
// like a find. Ask for a floor with Kinship.MinGrade rather than filtering after
// the fact — the engine withholds weaker rows, which is cheaper and honest.
//
// A [Corpus] is a scope, not a handle: it holds no resources and is safe to keep
// and share. Each verb runs one query through the runtime's ladder — the
// in-process analytic plane when it can, the certified `relate` binary otherwise.
package relate

import (
	"context"
	"path/filepath"
	"slices"
	"strings"

	"github.com/The-Billy-Company/irregex/bindings/go/analytic"
	"github.com/The-Billy-Company/irregex/bindings/go/runtime"
)

// Corpus is the scope every verb asks its question inside: the roots, and the
// directory relative paths resolve against.
type Corpus struct {
	roots []string
	dir   string
}

// Over scopes the compression verbs to roots (none = the index roots).
func Over(roots ...string) *Corpus { return &Corpus{roots: roots} }

// In sets the working directory the query resolves relative paths against.
func (c *Corpus) In(dir string) *Corpus {
	c.dir = dir
	return c
}

// Rows runs any analytic verb and hands back the undecoded cursor — the escape
// hatch for a caller that wants [runtime.Stats], a nested row this package's
// typed views flatten, or a verb added to the contract after this binding shipped.
func (c *Corpus) Rows(ctx context.Context, op analytic.Op, params analytic.Params) (*runtime.Rows, error) {
	return runtime.Run(ctx, runtime.Query{Op: op, Params: params, Roots: c.roots, Dir: c.dir})
}

// collect drains one verb into a typed slice. Stats are dropped here on purpose:
// a typed row list is the common case, and a caller who needs the answer-level
// counters (foreign, omitted) reaches for Rows instead.
func collect[T any](ctx context.Context, c *Corpus, op analytic.Op, params analytic.Params, scan func(runtime.Row) T) (out []T, err error) {
	rows, err := c.Rows(ctx, op, params)
	if err != nil {
		return nil, err
	}
	defer func() {
		if closeErr := rows.Close(); err == nil {
			err = closeErr
		}
	}()
	for row, err := range rows.All() {
		if err != nil {
			return out, err
		}
		out = append(out, scan(row))
	}
	return out, nil
}

// Region is a span of one file — a whole file when the unit is file, the
// enclosing function when it is not.
type Region struct {
	Path      string
	LineStart int64
	LineEnd   int64
	Headline  string
}

func region(r runtime.Row) Region {
	return Region{
		Path:      r.Text("path"),
		LineStart: r.Int("line_start"),
		LineEnd:   r.Int("line_end"),
		Headline:  r.Text("headline"),
	}
}

func regions(r runtime.Row, field string) []Region {
	nested := r.Rows(field)
	out := make([]Region, 0, len(nested))
	for _, child := range nested {
		out = append(out, region(child))
	}
	return out
}

// Neighbor is one ranked kinship answer: how far the probe is from this path, in
// which channel, and what that distance is worth.
type Neighbor struct {
	Path     string
	Distance float64
	Grade    analytic.Grade
	Channel  analytic.Channel
}

// Neighbors decodes the similar rows nested inside another answer (blast's
// twins), so a composed verb reuses this view rather than restating it.
func Neighbors(rows []runtime.Row) []Neighbor {
	out := make([]Neighbor, 0, len(rows))
	for _, row := range rows {
		out = append(out, neighbor(row))
	}
	return out
}

func neighbor(r runtime.Row) Neighbor {
	grade, _ := analytic.ParseGrade(r.Enum("grade").Label)
	channel, _ := analytic.ParseChannel(r.Enum("channel").Label)
	return Neighbor{Path: r.Text("path"), Distance: r.Float("distance"), Grade: grade, Channel: channel}
}

// Similar ranks the corpus against one probe — a path, `path#Lnnn` for the
// function containing that line, or bare text. The probe's shape decides how it
// is priced, which is why one verb covers "what resembles this file" and "what
// does this sentence sound like".
//
// A file is never its own kin: a distance-0 row for the probe itself is dropped,
// so the top row is the nearest OTHER thing regardless of which tier answered.
func (c *Corpus) Similar(ctx context.Context, k analytic.Kinship) ([]Neighbor, error) {
	near, err := collect(ctx, c, analytic.OpSimilar, k, neighbor)
	self := c.resolve(k.Target)
	if self == "" {
		return near, err
	}
	return slices.DeleteFunc(near, func(n Neighbor) bool { return c.resolve(n.Path) == self }), err
}

// resolve is a probe or row path's identity for comparison, "" when it does not
// name a file (bare text, or a `path#Lnnn` function probe, is not a whole unit and
// so cannot collide with one).
func (c *Corpus) resolve(path string) string {
	if path == "" || strings.Contains(path, "#L") {
		return ""
	}
	full := path
	if !filepath.IsAbs(full) {
		full = filepath.Join(c.dir, path)
	}
	resolved, err := filepath.EvalSymlinks(full)
	if err != nil {
		return ""
	}
	return resolved
}

// Pair is two paths close enough in bytes to be the same thing twice.
type Pair struct {
	A, B     string
	Distance float64
	Grade    analytic.Grade
}

// Dups are the near-duplicate pairs, closest first — copy-paste and its drift.
func (c *Corpus) Dups(ctx context.Context, k analytic.Kinship) ([]Pair, error) {
	return collect(ctx, c, analytic.OpDups, k, func(r runtime.Row) Pair {
		grade, _ := analytic.ParseGrade(r.Enum("grade").Label)
		return Pair{A: r.Text("a"), B: r.Text("b"), Distance: r.Float("distance"), Grade: grade}
	})
}

// Cluster is a fork family: the transitive closure of the verified duplicate
// graph, graded by its loosest edge.
type Cluster struct {
	Paths       []string
	MaxDistance float64
	Grade       analytic.Grade
}

// Clusters are the fork families, largest first — the restructure-ready unit a
// pair list makes a caller re-derive with its own union-find.
func (c *Corpus) Clusters(ctx context.Context, k analytic.Kinship) ([]Cluster, error) {
	return collect(ctx, c, analytic.OpClusters, k, func(r runtime.Row) Cluster {
		grade, _ := analytic.ParseGrade(r.Enum("grade").Label)
		return Cluster{Paths: r.Strings("paths"), MaxDistance: r.Float("max_distance"), Grade: grade}
	})
}

// Echo is a pair far apart in bytes but close in structure: the same skeleton
// wearing different vocabulary. Echo is that gap, and higher is stronger — the
// inverse polarity of a distance, which is why it has its own column.
type Echo struct {
	A, B      string
	Echo      float64
	Bytes     float64
	Structure float64
	Grade     analytic.Grade
}

// Echoes are the DRY candidates duplicate detection cannot see, widest gap first.
func (c *Corpus) Echoes(ctx context.Context, k analytic.Kinship) ([]Echo, error) {
	return collect(ctx, c, analytic.OpEchoes, k, func(r runtime.Row) Echo {
		grade, _ := analytic.ParseGrade(r.Enum("grade").Label)
		return Echo{
			A: r.Text("a"), B: r.Text("b"),
			Echo:      r.Float("echo"),
			Bytes:     r.Float("byte_distance"),
			Structure: r.Float("structure_distance"),
			Grade:     grade,
		}
	})
}

// Concept is one repeated implementation shape across several functions — the
// abstraction candidate. Bytes and Echo are optional because a family of
// structurally identical fragments may have no measurable byte kinship at all.
type Concept struct {
	Members       []Region
	RepeatedLines int64
	Confidence    float64
	Structure     float64
	Bytes         *float64
	Echo          *float64
}

// Concepts are the function-level shape families: what this corpus keeps
// re-implementing under different names.
func (c *Corpus) Concepts(ctx context.Context, k analytic.Kinship) ([]Concept, error) {
	return collect(ctx, c, analytic.OpConcepts, k, func(r runtime.Row) Concept {
		return Concept{
			Members:       regions(r, "members"),
			RepeatedLines: r.Int("repeated_lines"),
			Confidence:    r.Float("confidence"),
			Structure:     r.Float("structure_distance"),
			Bytes:         opt(r, "byte_distance"),
			Echo:          opt(r, "echo"),
		}
	})
}

// Family is one repetition family with its own rank, unit and channel — the
// generalized cluster, reused by the composed `family` verb.
type Family struct {
	Rank          int64
	Unit          analytic.Unit
	Channel       analytic.Channel
	Edge          float64
	RepeatedLines int64
	Score         float64
	Members       []Region
}

// ScanFamily decodes one family row, so the composed plane reuses this view.
func ScanFamily(r runtime.Row) Family {
	unit, _ := analytic.ParseUnit(r.Enum("unit").Label)
	channel, _ := analytic.ParseChannel(r.Enum("channel").Label)
	return Family{
		Rank:          r.Int("rank"),
		Unit:          unit,
		Channel:       channel,
		Edge:          r.Float("edge"),
		RepeatedLines: r.Int("repeated_lines"),
		Score:         r.Float("score"),
		Members:       regions(r, "members"),
	}
}

// Fragments are the function-level families: a twelve-line helper cloned into six
// files is a family here even where those files share three percent of their bytes.
func (c *Corpus) Fragments(ctx context.Context, k analytic.Kinship) ([]Family, error) {
	return collect(ctx, c, analytic.OpFragments, k, ScanFamily)
}

// Lone is a unit with no relative, priced by its nearest miss — the complement
// that turns "which of these fourteen implementations is genuinely unique" into a
// measurement.
type Lone struct {
	Unit      analytic.Unit
	Member    Region
	Nearest   *Region
	Bytes     float64
	Structure float64
}

// Distinct are the unrepeated units, each carrying the closest thing to a
// relative it has.
func (c *Corpus) Distinct(ctx context.Context, k analytic.Kinship) ([]Lone, error) {
	return collect(ctx, c, analytic.OpDistinct, k, func(r runtime.Row) Lone {
		unit, _ := analytic.ParseUnit(r.Enum("unit").Label)
		l := Lone{
			Unit:      unit,
			Bytes:     r.Float("byte_distance"),
			Structure: r.Float("structure_distance"),
		}
		if members := r.Rows("member"); len(members) > 0 {
			l.Member = region(members[0])
		}
		if nearest := r.Rows("nearest"); len(nearest) > 0 {
			near := region(nearest[0])
			l.Nearest = &near
		}
		return l
	})
}

// opt reads a measurement that may not have been taken. A zero distance means
// identical, so absence cannot be spelled as zero.
func opt(r runtime.Row, field string) *float64 {
	v, ok := r.OptFloat(field)
	if !ok {
		return nil
	}
	return &v
}
